Lib/typing.py

```diff
@@ -177,7 +179,8 @@ def _type_check(arg, msg, is_argument=True, module=None, *, allow_special_forms=
         return arg
     if isinstance(arg, _SpecialForm) or arg in (Generic, Protocol):
         raise TypeError(f"Plain {arg} is not valid as type argument")
-    if isinstance(arg, (type, TypeVar, ForwardRef, types.UnionType, ParamSpec)):
+    if isinstance(arg, (type, TypeVar, TypeVarTuple, ForwardRef,
```

Comment:
`_type_check` seems to be used in two kinds of places:

1. `_type_check` is used on the type arguments for `SpecialGenericAlias`.
2. Apart from the above, `_type_check` is used on the type arguments of a bunch of special forms like `Optional`, `Final`, `Union`, etc.

IIUC, we care about variadic tuples in (1) but not (2). 

For (1), if we have `MyAlias = Array[*Ts]`, we want to be able to instantiate the alias as `MyAlias[int, *Ts2, str]`. That suggests what we want above is `Unpack`.

For (2), with the current code, we'd accept `x: Final[Ts]` or `Optional[Ts]`, which again isn't valid. Note that even `Optional[*Ts]` is wrong. Likewise, `BoundVarianceMixin` uses `_type_check` on the bound. This means we'd end up accepting `T = TypeVar("T", bound=Ts)`.

(I haven't tested the above, so I could be way off.)

I see two ways ahead:
(a) replace `TypeVarTuple` with `Unpack` above and just live with the invalid uses like `Optional[*Ts]`.
(b) make `SpecialGenericAlias` check for `Unpack` on top of `_type_check` so that it can accept `MyAlias[int, *Ts, str]` but things like `Optional[*Ts]` would be ruled out.

What do you think? If that sounds right, could you add tests for these cases?

Reply:
Having thought about this for a couple of days, I lean towards (a). Both ways I can think of implementing (b) seem bad:
* Implementation 1: Leave `_type_check` unmodified, but add extra code to `SpecialGenericAlias` to make it ok with taking TypeVarTuples. This seems ugly because i) it means that `_type_check` is no longer the central source of truth about whether something's a valid type argument, and ii) it introduces potential maintenance overhead (if we need to change it in the future, we need to remember to change it in all the `SpecialGenericAlias`-like places - e.g. currently `_type_check` is used both in `_SpecialGenericAlias.__getitem__` and `_TupleType.__getitem__`).
* Implementation 2: Modify `_type_check` to give it an extra argument controlling whether it's ok with `TypeVarTuple`. I hesitate to do this because `_type_check` already feels like a bit of a mess - I don't want to make it any more complex.

A second consideration is that the general vibe I've got from folks is that they'd prefer to keep the restrictions on typing at runtime relatively minimal - e.g. what you were saying about preferring to allow bound and variance arguments to `TypeVarTuple` at runtime, Jelle's argument that keeping restrictions minimal at runtime allows people to experiment. I'm generally in favour of this  - at least when it doesn't lead to other inconsistencies.

So yeah, overall I think we should live with the badness of allowing e.g. `Optional[*Ts]` and `Final[*Ts]` - I think it's worth the savings in complexity.

I'll resolve this for now, but do re-open if you want to discuss it more/can think of a better way of implementing (b).